src/main/scala/chisel3/util/experimental/BarrelShifter.scala

```diff
@@ -0,0 +1,80 @@
+  * @param shiftInput       input signal to indicate the shift number, encoded in UInt.
+  * @param shiftType        [[ShiftType]] to indicate the type of shifter.
+  * @param shiftGranularity how many bits will be resolved in each layer.
+  *                         For a smaller `shiftGranularity`, latency will be high, but area is smaller.
+  *                         For a large `shiftGranularity`, latency will be low, but area is higher.
+  *
+  * @example {{{
+  * val input = 0xAABBCCDD.U(32.W)
+  * val bytes = VecInit(input.asBools.grouped(8).map(VecInit(_).asUInt).toSeq)
+  * val shamt = 1.U(2.W)
+  * val output = BarrelShifter.leftRotate(bytes, shamt).asUInt // big endian
+  * // output = 0xDDAABBCC.U(32.W)
+  * // The depth of Mux is 2, whereas originally
+  * // the depth of Mux is log(32) = 5
+  * val output = BarrelShifter.leftRotate(bytes, shamt, 2).asUInt // depth is 1 now
+  * }}}
+  */
+object BarrelShifter {
+  trait ShiftType
+
+  object LeftShift extends ShiftType
+
+  object RightShift extends ShiftType
+
+  object LeftRotate extends ShiftType
+
+  object RightRotate extends ShiftType
+
+  def apply[T <: Data](inputs: Vec[T], shiftInput: UInt, shiftType: ShiftType, shiftGranularity: Int = 1): Vec[T] = {
+    val elementType: T = chiselTypeOf(inputs.head)
+    shiftInput
+      .asBools()
+      .grouped(shiftGranularity)
+      .map(VecInit(_).asUInt())
+      .zipWithIndex
+      .foldLeft(inputs) {
+        case (prev, (shiftBits, layer)) =>
+          Mux1H(
+            UIntToOH(shiftBits),
+            Seq.tabulate(1 << shiftBits.getWidth)(i => {
+              // shift no more than inputs length
+              // prev.drop will not warn about overflow!
+              val layerShift: Int = (i * (1 << (layer * shiftGranularity))).min(prev.length)
+              VecInit(shiftType match {
+                case LeftRotate =>
+                  prev.drop(layerShift) ++ prev.take(layerShift)
+                case LeftShift =>
+                  prev.drop(layerShift) ++ Seq.fill(layerShift)(0.U.asTypeOf(elementType))
+                case RightRotate =>
+                  prev.takeRight(layerShift) ++ prev.dropRight(layerShift)
+                case RightShift =>
+                  Seq.fill(layerShift)(0.U.asTypeOf(elementType)) ++ prev.dropRight(layerShift)
+              })
+            })
+          )
+      }
+  }
+
+  def leftShift[T <: Data](inputs: Vec[T], shift: UInt, layerSize: Int = 1): Vec[T] =
+    apply(inputs, shift, LeftShift, layerSize)
+
+  def rightShift[T <: Data](inputs: Vec[T], shift: UInt, layerSize: Int = 1): Vec[T] =
+    apply(inputs, shift, RightShift, layerSize)
+
+  def leftRotate[T <: Data](inputs: Vec[T], shift: UInt, layerSize: Int = 1): Vec[T] =
+    apply(inputs, shift, LeftRotate, layerSize)
+
+  def rightRotate[T <: Data](inputs: Vec[T], shift: UInt, layerSize: Int = 1): Vec[T] =
+    apply(inputs, shift, RightRotate, layerSize)
```

Comment:
Need documentations on these public APIs.

Reply:
The doc for them is the doc of the object. I think the `@param` and `@example` above is quite clear for users.